### PR TITLE
Some thoughts on extensions (+ a straight-up typo)

### PR DESCRIPTION
**Pull request by goldfirere**
I've taken a pass through your (very helpful) listing of extensions and added my own thoughts.

Before looking at the changes to `extensions.html`, make sure to see the changes in `tutorial.md`, where I tighten the definition of some of the columns. Happy to debate any of these changes.

### CODE REVIEW

**Comment by sdiehl** [comment]
Hi Richard,

Thanks for looking at this section and fixing the typos. I like your extension to the definition of Benign me rethink some of my old assumptions for this section. To summarize the changes that I agree with:

AutoDeriveTypable -> Benign :white_check_mark: 
ConstrainedClassMethods -> Benign :white_check_mark: 
EmptyCase -> Benign :white_check_mark: 
ExplicitForall -> Steals Syntax :white_check_mark: 
ExtendedDefaultRules -> Benign :white_check_mark: 
FlexibleContexts -> Benign  :white_check_mark: 
FlexibleInstances -> Benign  :white_check_mark: 
GADTSyntax -> not(Steals Syntax) :white_check_mark: 
GeneralizedNewtypeDeriving -> Benign :white_check_mark: 
ImpredicativeTypes -> Historical :white_check_mark: 
KindSignatures -> Benign :white_check_mark: 
LiberalTypeSynonyms -> Benign  :white_check_mark: 
MagicHash -> Benign  :white_check_mark: 
MultiWayIf  ->  not(Steals Syntax) :white_check_mark: 
NamedFieldPuns -> Benign :white_check_mark: 
ParallelListComp -> Historical :white_check_mark: 
PatternGuards -> Benign :white_check_mark: 
RecursiveDo -> Benign  :white_check_mark: 
RoleAnnotations -> Benign :white_check_mark: 
TypeApplications -> Benign :white_check_mark: 
TypeApplications -> Steals Syntax :white_check_mark: 
TypeFamilies -> Benign :white_check_mark: 
TypeOperators -> Benign  :white_check_mark: 
UnboxedTuples -> Benign :white_check_mark: 
StealsSyntax -> Benign :white_check_mark: 
ConstraintKinds -> Benign :white_check_mark: 
ImplicitParams -> Benign :white_check_mark: 

To summarize the ones I'm still on the fence on, I consider that enabling these extension do sometimes cause inference and error reporting to change within the module which makes them "not benign". While they don't change the semantics, compiling may be effected by enabling these extensions. Open to debate on this topic.

Cpp -> Benign 
GADTs -> Benign
ExistentialQuantification -> Benign
ScopedTypeVariables -> Benign

**Comment by goldfirere** [comment]
I changed `CPP` to be not benign because the C preprocessor sometimes chokes on valid Haskell code. For example, ending a line with a `\` is a problem in a module with `CPP` enabled. (I discovered this after some pain in assigning an operator `\\` a fixity in a module with `CPP` enabled. The fix was to include a comment late on the same line. Very annoying.)

Enabling `GADTs` enables `MonoLocalBinds` which does change type inference behavior. Perhaps I agree that this makes `GADTs` non-benign. `TypeFamilies` also enables `MonoLocalBinds`.

I don't see any problem with `ExistentialQuantification`. What are you concerned about?

Upon further thought, I agree that `ScopedTypeVariables` is non-benign if enabled in the presence of `ExplicitForAll`, as `ScopedTypeVariables` would `forall`-bound variables newly into scope.

I wonder if a new column "Generally acceptable" is warranted that would include any extension that doesn't cause real trouble. For example I agree that `GADTs` is not benign according to the written definition (it can break previously-accepted programs via enabling `MonoLocalBinds`), but no one would look askance at a module with `GADTs` enabled. Contrast to `IncoherentInstances`.

**Comment by sdiehl** [comment]
I think you're right about ExistentialQuantification, that shouldn't adversely effect anyone accidentally be enabling it.

As for what's generally acceptable, I think I won't try to address that in the table other than to indicate whether it's specialized or general.. There's a "The Dangerous" section that indicates you probably don't want IncoherentInstances or ImpredicativeTypes.

Otherwise I think this all looks good and I'm happy to merge it now. Thanks for the feedback.
